### PR TITLE
Added support for getting configuration from a config file

### PR DESCRIPTION
**Pull request by l4zy0n3**
1. Changed argument parsing to use argparse
2. Fixed, #363 
3. Added some temporary workarounds when loading from config
#352

### CODE REVIEW

File: examples/guided.py [hunk 1]
```diff
@@ -388,5 +387,17 @@ def perform_installation(mountpoint):
 			except:
 				pass
 
-ask_user_questions()
+if not archinstall.arguments:
```
Comment (dylanmtaylor) [inline examples/guided.py:390]: Hmm, I feel this is too broad of a condition and might restrict what we can do too much. For instance, what if we want to do --config but have the user just provide passwords so they are not stored in a config file? Being able to selectively ask questions for just things not provided could be a valuable feature in the future.

**Comment by dylanmtaylor** [comment]
Having nearly 30 commits for a relatively small change like this makes it hard to follow what's going on, perhaps consider squashing them to make history easier to follow in the future before it's merged.

**Comment by dylanmtaylor** [comment]
Also finding it really weird there are no built ISOs from these changes - do you have actions enabled on your fork?

**Comment by l4zy0n3** [comment]
> Also finding it really weird there are no built ISOs from these changes - do you have actions enabled on your fork?

No I don't, sorry I'm relatively new to this, will do asap

**Comment by dylanmtaylor** [comment]
> > Also finding it really weird there are no built ISOs from these changes - do you have actions enabled on your fork?
> 
> No I don't, sorry I'm relatively new to this, will do asap

So in theory you shouldn't even need to have it enabled - we want to have every PR get built so we can test them before merging. I'm not sure why yours are not building...

File: examples/guided.py (hunk 1 above)
Comment (l4zy0n3) [reply examples/guided.py:390]: That should be the proper way to implement this, I was kind of going only for the basic ( silent ) install for now, so we could have declarative configs and everything else is automated... I might be able to modify this in the future to cover broader perspectives

**Comment by l4zy0n3** [comment]
> > > Also finding it really weird there are no built ISOs from these changes - do you have actions enabled on your fork?
> > 
> > 
> > No I don't, sorry I'm relatively new to this, will do asap
> 
> So in theory you shouldn't even need to have it enabled - we want to have every PR get built so we can test them before merging. I'm not sure why yours are not building...

image
I think it needs to be approved first

File: examples/guided.py (hunk 1 above)
Comment (l4zy0n3) [reply examples/guided.py:390]: Would it be fine if the user sets sensitive info such as passwords through environment variables?

**Comment by dylanmtaylor** [comment]
> > > > Also finding it really weird there are no built ISOs from these changes - do you have actions enabled on your fork?
> > > 
> > > 
> > > No I don't, sorry I'm relatively new to this, will do asap
> > 
> > 
> > So in theory you shouldn't even need to have it enabled - we want to have every PR get built so we can test them before merging. I'm not sure why yours are not building...
> 
> image
> I think it needs to be approved first

@Torxed would have to approve that

File: examples/guided.py (hunk 1 above)
Comment (dylanmtaylor) [reply examples/guided.py:390]: Hmm, that could definitely be viable. It'd be strange to do as there is not a fixed number of users though. You can have as many as you want in ArchInstall.

**Comment by l4zy0n3** [comment]
Hi @Torxed, is this still applicable or shall I close this

**Comment by Torxed** [comment]
> Hi @Torxed, is this still applicable or shall I close this

Keep it open, I simply haven't had time to test this yet to fully understand the changes involved.

**Comment by dylanmtaylor** [comment]
@l4zy0n3 There are a LOT of commits in this request, could you do a rebase and squash them into a "Add support for getting configuration from a config file" or maybe 3-4 commits? Having >30 commits for a 46 line change makes it really hard to follow history in the project.

**Comment by l4zy0n3** [comment]
> @l4zy0n3 There are a LOT of commits in this request, could you do a rebase and squash them into a "Add support for getting configuration from a config file" or maybe 3-4 commits? Having >30 commits for a 46 line change makes it really hard to follow history in the project.

Yes, i'm working on it

**Comment by dylanmtaylor** [comment]
I really like this change. Before this is merged, I'd also like to see an example configuration file and a write-up of how to use the example file with archinstall added to the documentation. I think that would be greatly beneficial to users and contributors alike.

**Comment by l4zy0n3** [comment]
> I really like this change. Before this is merged, I'd also like to see an example configuration file and a write-up of how to use the example file with archinstall added to the documentation. I think that would be greatly beneficial to users and contributors alike.

Hi @dylanmtaylor, currently i'm facing #417 post install, i'm able to successfully chroot into the installation though, I will add a config and a writeup of some sort while waiting for this to be resolved

File: examples/guided.py (hunk 1 above)
Comment (l4zy0n3) [reply examples/guided.py:389]: Can we have a .env file to pull users and passwords from, currently testing this locally and it seems to be working, but this will involve adding a requirement (**dotenv**)

File: pyproject.toml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 [build-system]
-requires = ["flit_core >=2,<4"]
+requires = ["flit_core >=2,<4", "python-dotenv"]
```
Comment (dylanmtaylor) [inline pyproject.toml:2]: Should this be version pinned?

File: .github/workflows/python-publish.yml [hunk 3]
```diff
@@ -27,4 +27,4 @@ jobs:
           FLIT_USERNAME: ${{ secrets.PYPI_USERNAME }}
           FLIT_PASSWORD: ${{ secrets.PYPI_PASSWORD }}
         run: |
-          flit publish
+          flit publish
```
Comment (dylanmtaylor) [inline .github/workflows/python-publish.yml:30]: not sure why this file is even touched in this PR, not a big deal, just a curiosity.

File: examples/.env [hunk 4]
```diff
@@ -0,0 +1,3 @@
+    ROOT_PASSWD=<root_password>
```
Comment (dylanmtaylor) [inline examples/.env:1]: Can you put this line in quotes so it's technically valid like the other ones?

Comment (l4zy0n3) [reply examples/.env:1]: Sure

File: .github/workflows/python-publish.yml (hunk 3 above)
Comment (l4zy0n3) [reply .github/workflows/python-publish.yml:30]: Sure, will clean this up

File: pyproject.toml (hunk 2 above)
Comment (l4zy0n3) [reply pyproject.toml:2]: I will remove this in the latest commit as I already have defined this dependency in setup.cfg

File: pyproject.toml [hunk 5]
```diff
@@ -27,4 +27,4 @@ include = ["docs/","profiles"]
 exclude = ["docs/*.html", "docs/_static","docs/*.png","docs/*.psd"]
 
 [tool.flit.metadata.requires-extra]
-doc = ["sphinx"]
```
Comment (l4zy0n3) [inline pyproject.toml:30]: Seems like commiting from github automatically adds a newline at the end

**Comment by dylanmtaylor** [comment]
Made a PR to merge master into this here https://github.com/n1x0ps/archinstall/pull/4

**Comment by dylanmtaylor** [comment]
Odd that I'm seeing merge conflicts since there was a merge commit in that PR, let me try to fix it...

File: archinstall/lib/systemd.py [hunk 6]
```diff
@@ -103,7 +104,7 @@ def is_alive(self):
 
 		return self.session.is_alive()
 
-	def SysCommand(self, cmd :list, *args, **kwargs):
+def SysCommand(self, cmd: list, *args, **kwargs):
```
Comment (Torxed) [inline archinstall/lib/systemd.py:107]: @dylanmtaylor Not sure what's going on here, the function is a class function so it shouldn't be indented in to the left.

**Review by Torxed — CHANGES_REQUESTED** [review]

File: archinstall/lib/systemd.py (hunk 6 above)
Comment (l4zy0n3) [reply archinstall/lib/systemd.py:107]: my bad, I messed up while resolving conflicts

Comment (dylanmtaylor) [reply archinstall/lib/systemd.py:107]: Should be an easy fix then, just need to fix indentation. flake8 caught this too.

./archinstall/lib/systemd.py:1:1: E902 IndentationError: unindent does not match any outer indentation level
1     E902 IndentationError: unindent does not match any outer indentation level

**Comment by dylanmtaylor** [comment]
I think it still needs that indent fix.

File: archinstall/lib/systemd.py (hunk 6 above)
Comment (Torxed) [reply archinstall/lib/systemd.py]: Ah, sorry. Thought it was something from the merge PR.

Comment (l4zy0n3) [reply archinstall/lib/systemd.py]: Okay, fixed the indent and removed a redundant import, can you guys verify

**Comment by l4zy0n3** [comment]
> I think it still needs that indent fix.

done that, we're back to 30+ commits 😂

**Comment by Torxed** [comment]
> > I think it still needs that indent fix.
> 
> done that, we're back to 30+ commits 😂

That's ok, I usually have 200 commits in my PR's :) Which is why I some times squash them during merge ^^

File: archinstall/lib/networking.py [hunk 7]
```diff
@@ -1,4 +1,5 @@
 import fcntl
+import logging
 import os
 import logging
```
Comment (Torxed) [inline archinstall/lib/networking.py:4]: We do have duplicate imports here tho, not sure this is needed. Same with `log` below.

**Review by Torxed — CHANGES_REQUESTED** [review]

File: archinstall/lib/networking.py (hunk 7 above)
Comment (l4zy0n3) [reply archinstall/lib/networking.py]: fixed

**Comment by Torxed** [comment]
Looks good to me, ready to go in once the ISO has been built on this PR and I can run through one installation at least.

File: setup.cfg [hunk 8]
```diff
@@ -22,7 +22,7 @@ classifers =
 [options]
 packages = find:
 python_requires = >= 3.8
-
+
```
Comment (dylanmtaylor) [inline setup.cfg:25]: Super minor - can you undo this whitespace change?

**Comment by Torxed** [comment]
image

Installs and boots just fine! :) Good work on this one!

File: setup.cfg (hunk 8 above)
Comment (l4zy0n3) [reply setup.cfg:25]: Sorry, missed your comment
